inspectit-oce-core/src/main/java/rocks/inspectit/oce/core/instrumentation/NewClassDiscoveryService.java

```diff
@@ -0,0 +1,79 @@
+package rocks.inspectit.oce.core.instrumentation;
+
+import lombok.extern.slf4j.Slf4j;
+import lombok.val;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Service;
+import rocks.inspectit.oce.core.config.InspectitEnvironment;
+import rocks.inspectit.oce.core.utils.StopWatch;
+
+import javax.annotation.PostConstruct;
+import java.lang.instrument.Instrumentation;
+import java.time.Duration;
+import java.util.*;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.TimeUnit;
+import java.util.function.Consumer;
+
+@Service
+@Slf4j
+public class NewClassDiscoveryService {
+
+    @Autowired
+    private InspectitEnvironment env;
+
+    @Autowired
+    private AsyncClassTransformer transformer;
+
+    @Autowired
+    private Instrumentation instrumentation;
+
+    @Autowired
+    private ScheduledExecutorService exec;
+
+
+    private Set<Class<?>> knownClasses = Collections.newSetFromMap(new WeakHashMap<>());
+
+    private Set<Consumer<Set<Class<?>>>> classDiscoveryListeners = Collections.newSetFromMap(new ConcurrentHashMap<>());
+
+    private Runnable updateCheckTask = () -> {
+        long timeSinceLastClassDefinition = System.currentTimeMillis() - transformer.getLastNewClassDefinitionTimestamp().get();
+        if (timeSinceLastClassDefinition < env.getCurrentConfig().getInstrumentation().getInternal().getMaxClassDefinitionDelay().toMillis()) {
+            val watch = new StopWatch();
+
+            log.debug("Checking for new classes...");
+            Set<Class<?>> newClasses = new HashSet<>();
+            for (Class<?> clazz : instrumentation.getAllLoadedClasses()) {
+                if (!knownClasses.contains(clazz)) {
+                    knownClasses.add(clazz);
+                    newClasses.add(clazz);
+                }
+            }
+            if (!newClasses.isEmpty()) {
+                log.debug("{} new classes found, check took {} ms (last class define {} ms ago)", newClasses.size(), watch.getElapsedMillis(), timeSinceLastClassDefinition);
+                classDiscoveryListeners.forEach(lis -> lis.accept(newClasses));
+            } else {
+                log.debug("No new classes found  check took {} ms (last class define {} ms ago)", watch.getElapsedMillis(), timeSinceLastClassDefinition);
+            }
+        }
+        scheduleUpdateCheck();
+    };
+
+
+    public void addDiscoveryListener(Consumer<Set<Class<?>>> listener) {
```

Comment:
same as for the `AsyncListener` I would create the interface and use auto-wire for the listeners.. The benefit would be that you could already pass the known classes on the `init()`, this way the other component does not have to do it as well and should only listen to the updates..